Add 'reflect_invoke' metafunction and motivating example.

katzdm opened the pull request:
I've discussed this one for a while with @daveedvdv , and we've both had success implementing it. He had no objections, suggested that I write up a motivating example, and propose the wording - so this is that :)

SG7 expressed interest in typeful reflection in Kona, so it's possible the example could help win some new fans to Team 2996.

brevzin commented on 2996_reflection/reflection.md:
````diff
@@ -996,6 +997,86 @@ Thus, the only instantiations of `TU_Ticket::Helper` occur because of the call t
+```cpp
+// Represents a 'std::meta::info' constrained by a predicate.
+template <std::meta::info Pred>
+  requires (type_of(reflect_invoke(Pred, {reflect_value(^int)})) == ^bool)
````
Why do you need `reflect_invoke` here if `Pred` is already a constant? Can't this just be `decltype([: Pred:](^int))`?

brevzin commented on 2996_reflection/reflection.md:
````diff
@@ -996,6 +997,86 @@ Thus, the only instantiations of `TU_Ticket::Helper` occur because of the call t
+  Printer::PrintKind([:enrich(^main):]);  // "function"
+  Printer::PrintKind([:enrich(^3):]);     // "unknown kind"
+}
+```
````
Missing `:::` here.

brevzin commented on 2996_reflection/reflection.md:
```diff
@@ -996,6 +997,86 @@ Thus, the only instantiations of `TU_Ticket::Helper` occur because of the call t
+template <typename... Choices>
+constexpr auto enrich = [](std::meta::info r) consteval {
+  std::array checks = {^Choices::check...};
+  std::array ctors = {^Choices::make...};
```
Is there a way that we can just use the constructors? Since that's really all the `make` functions are here for.

katzdm replied:
Done.

brevzin commented on 2996_reflection/reflection.md:
```diff
@@ -996,6 +997,86 @@ Thus, the only instantiations of `TU_Ticket::Helper` occur because of the call t
+  requires (type_of(reflect_invoke(Pred, {reflect_value(^int)})) == ^bool)
+struct metatype {
+  // The underlying 'std::meta::info' value.
+  const std::meta::info value;
```
```suggestion
  std::meta::info value;
```

brevzin commented on 2996_reflection/reflection.md:
```diff
@@ -1660,6 +1744,22 @@ namespace std::meta {
+
+Letting `F` be the entity reflected by `target`, and `A_0, ..., A_n` be the sequence of entities reflected by the values held by `args`: if the expression `F(A_0, ..., A_N)` is a well-formed constant expression evaluating to a type that is not `void`, and if every value in `args` is a reflection of a constant value, then `reflect_invoke(target, args)` evaluates to a reflection of the constant val ...
+
+For all other invocations, `reflect_invoke(target, args)` is ill-formed.
```
I think we should put `substitute` and `reflect_invoke` next to each other, since they're pretty similar - substitute is `^[: templ :]<[: args :]...>` and `reflect_invoke` is `^[: f :]([: args :]...)`. 

Should also add an example or two.

brevzin commented on 2996_reflection/reflection.md:
```diff
@@ -996,6 +997,86 @@ Thus, the only instantiations of `TU_Ticket::Helper` occur because of the call t
+  // Demonstration of using 'enrich' to select an overload.
+  Printer::PrintKind([:enrich(^int):]);   // "type"
+  Printer::PrintKind([:enrich(^main):]);  // "function"
+  Printer::PrintKind([:enrich(^3):]);     // "unknown kind"
```
There's no `Printer` in this example right now.

katzdm commented:
@brevzin Addressed all feedback. Did it with constructors, as per your request - I _think_ the existing language of the wording is sufficient to cover that, but let me know if you'd prefer a more explicit mention that calling constructors is fair game.

brevzin commented on 2996_reflection/reflection.md:
```diff
@@ -999,6 +1000,87 @@ Thus, the only instantiations of `TU_Ticket::Helper` occur because of the call t
+template <typename... Choices>
+consteval std::meta::info enrich(std::meta::info r) {
+  std::array ctors = {members_of(^Choices, std::meta::is_constructor)[0]...,
+                      members_of(^unmatched, std::meta::is_constructor)[0]};
```
Why only taking the first constructor? I guess what you need is the constructor that takes a `std::meta::info`, but this relies on that one being ahead of the copy and move constructors?

katzdm replied:
Well I control the class, and the Standard guarantees that implicitly-declared special member functions are considered to be declared at the closing `}` of the class definition:
https://eel.is/c++draft/class#special

> An implicitly-declared special member function is declared at the closing } of the class-specifier.

And `members_of` guarantees iteration in declaration order, so for the purpose of this example, this seems safe to me.
